shark_turbine/kernel/wave/thread_shape_analysis.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from ...support.logging import get_logger
+from shark_turbine.kernel._support.tracing import CapturedTrace
+import torch.fx as fx
+from ..ops.wave_ops import *
+from ..lang.global_symbols import *
+from .utils import capture_forward_slice, capture_backward_slice
+
+logger = get_logger("turbine.wave.thread_shape_analysis")
+
+
+@dataclass(order=True)
+class IndexSize:
+    index: IndexSymbol
+    size: int
+
+    def __hash__(self):
+        return hash((self.index, self.size))
+
+
+def get_index_sizes(indices: list[IndexSequence]):
+    dims = frozenset([IndexSize(index, seq.size) for index, seq in indices.items()])
+    return dims
+
+
+def get_custom_index_sizes(custom: CustomOp):
+    return get_index_sizes(custom.index)
+
+
+def set_index_size(custom: CustomOp, target_index_sizes: list[IndexSize]):
+    for target in target_index_sizes:
+        if target.index not in custom.index:
+            raise NotImplementedError(
+                "NYI: Handle when source target index size is not found in target/user index."
+            )
+        custom.index[target.index].size = target.size
+
+
+# Function called on op post propagation for extra processing/handling.
+def post_propagation(custom: CustomOp, target_index_sizes: list[IndexSize]):
```

Comment:
Why is this function required? Don't the forward and backward slices go beyond the boundaries of the reduction and propagate from iter args to init args?

Reply:
backward slice does not go through from IterArg to init args. Shall we modify getInputs to do that instead(it's quite doable)? Since I am not the only user of bfs/backwardSlice I'd like to be mindful of modifiying too much  haha

Reply:
@harsh-nod I modified it, but not 100% sure this would work for other's use cases, would probably want your eyes on this one :)

Reply:
Yes I think that makes sense. Looks good to me!